### PR TITLE
Optimize Gem::Package::TarReader#each

### PR DESCRIPTION
**Pull request by casperisfine**
The current implementation always try to seek, and then rescue `NameError` (parent of `NoMethodError`) to fallback to seeking using `read`.

The main issue is that it never remembers that the IO don't support seeking, so it uselessly try again on each iteration. If you have 1k entries in a `tar.gz` it will raise `1k` exception with their backtraces etc.

A simple early check using `respond_to?(:seek)` can avoid all that.

I benched this patch against a `165M` linux tar.gz, and saw a `3.6%` improvement. [The benchmark code](https://gist.github.com/casperisfine/53171c14c135bf61979f328926d9ee34).

cc @rafaelfranca @chrisseaton @XrXr

### CODE REVIEW

File: lib/rubygems/package/tar_reader.rb [hunk 1]
```diff
@@ -67,18 +69,22 @@ def each
...
+      if use_seek
+        begin
+          # avoid reading if the @io supports seeking
+          @io.seek pending, IO::SEEK_CUR
+          pending = 0
+        rescue Errno::EINVAL
         end
       end
 
+      # if seeking didn't work
```
Comment (deivid-rodriguez) [inline lib/rubygems/package/tar_reader.rb]: I guess "if seeking is not supported" is more adequate here, the current wording seems more appropriate for the old way of rescuing the `NameError`?

Comment (casperisfine) [reply lib/rubygems/package/tar_reader.rb]: Well, It could also be caused by the `rescue Errno::EINVAL`, but you're right, I'll clarify.

Comment (deivid-rodriguez) [reply lib/rubygems/package/tar_reader.rb]: Indeed, I missed that code path :+1:

**Comment by bronzdoc** [comment]
@bundlerbot r+

**Comment by hsbt** [comment]
@bundlerbot r+

**Pull request merged**